pkg/image/ext3_test.go

```diff
@@ -0,0 +1,249 @@
+		t.Skipf("%s not available, skipping the test", cmdBin)
+	}
+
+	var out, err bytes.Buffer
+	cmd := exec.Command(cmdBin, "-F", "-t", fsType, path)
+	cmd.Stderr = &err
+	cmd.Stdout = &out
+	cmdErr := cmd.Run()
+	if cmdErr != nil {
+		t.Fatalf("command failed: %s - stderr: %s - stdout: %s\n", cmdErr, err.String(), out.String())
+	}
+}
+
+// createFullVirtualBlockDevice creates a full virtual
+// block device with a file system in it.
+// Warning the file will be earased!!!!
+// @param[in] test handle to control the test execution
+// @param[in] path to the virtual block device to be created.
+// @param[in] type of the file system to be created in the virtual block device (e.g., "ext3")
+func createFullVirtualBlockDevice(t *testing.T, path string, fsType string) {
+	createVirtualBlockDevice(t, path)
+	createFS(t, fsType, path)
+}
+
+// ext3InitializerTest prepares the initializer test by
+// creating a new virtual block device that will be
+// associated with the image, opening the image, get stat
+// information about the file associated to the image and
+// finally calling the initializer.
+// @param[in] test handle to control the test execution.
+// @param[in] image handle, when calling, simply an initialized Image variable.
+// @param[in] path to the file that will be used with the image; it does not need to exist when calling the function.
+// @return the error handle returned by the initializer. We do *not* intend to analyze the handle in this function since only the caller knows if it is a valid or invalid test case
+func ext3InitializerTest(t *testing.T, img *Image, path string, fsType string) error {
+	createFullVirtualBlockDevice(t, path, fsType)
+
+	var err error
+	img.File, err = os.Open(path)
+	if err != nil {
+		t.Fatalf("cannot open file: %s\n", err)
+	}
+
+	fileinfo, err := img.File.Stat()
+	if err != nil {
+		t.Fatalf("cannot stat image: %s\n", err)
+	}
+
+	var ext3format ext3Format
+	err = ext3format.initializer(img, fileinfo)
+	// err is just to be returned and analyzed by the caller
+
+	img.File.Close()
+
+	return err
+}
+
+func TestCheckExt3Header(t *testing.T) {
+	b := make([]byte, bufferSize)
+
+	// Create a fake ext3 file
+	dir, err := ioutil.TempDir("", "headerTesting-")
+	if err != nil {
+		t.Fatalf("impossible to create temporary directory: %s\n", err)
+	}
+	defer os.RemoveAll(dir)
+
+	path := dir + "ext3.fs"
+
+	createFullVirtualBlockDevice(t, path, "ext3")
+
+	// Now load the image
+	img, imgErr := os.Open(path)
+	if imgErr != nil {
+		t.Fatal("impossible to load image for testing")
+	}
+	defer img.Close()
+
+	n, err := img.Read(b)
+	if err != nil || n != bufferSize {
+		t.Fatalf("cannot read the first %d bytes of the image\n", bufferSize)
+	}
+
+	_, checkErr := CheckExt3Header(b)
+	if checkErr != nil {
+		t.Fatalf("cannot check ext3 header of a valid image: %s\n", checkErr)
+	}
+}
+
+func TestInitializer(t *testing.T) {
+	// Create a temporary image which is obviously an invalid ext3 image
```

Comment:
Tests not requiring privileges must drop privileges at the beginning of the function with snippet :

```
test.DropPrivilege(t)
defer test.ResetPrivilege(t)
```

Otherwise they could be executed as root, because `go test` is called with sudo, so all tests function must explicitly tell if they need to keep privileges or drop privileges during execution

For functions requiring privileges, a call `test.EnsurePrivilege(t)` at the beginning of the function is required

Reply:
Good to know! Should we audit the tests and make sure we actually do that? I do not recall seeing this since I started working on testing. What do you think?

Reply:
I revisit the entire code to rely on this. These tests should now be good to go.

Reply:
To be consistent yes, will open an issue and add it to epic